### PR TITLE
Automatic setup of on-device dependencies for iOS

### PR DESCRIPTION
**Pull request by zner0L**
This automatically installs missing dependencies on a jailbroken iOS device via `apt`.

### CODE REVIEW

File: src/ios.ts [hunk 1]
```diff
@@ -178,6 +213,8 @@ export const iosApi = <RunTarget extends SupportedRunTarget<'ios'>>(
             } catch (err) {
                 throw new Error('Cannot connect using SSH.', { cause: err });
             }
+
+            await this._internal.setupEnvironment();
```
Comment (baltpeter) [inline src/ios.ts:231]: You still have to update the README. As this only happens when the `ssh` capability is enabled and considering https://github.com/tweaselORG/appstraction/issues/59#issuecomment-1545720609, I guess you could just add a "These are installed automatically if…" note to the list of packages for now.

**Review by baltpeter — CHANGES_REQUESTED** [review]
Haven't tested this yet but a few comments from reading the code.

File: src/ios.ts [hunk 2]
```diff
@@ -145,6 +145,41 @@ export const iosApi = <RunTarget extends SupportedRunTarget<'ios'>>(
             ssh.dispose();
             return res;
         },
+        async setupEnvironment() {
+            if (!options.capabilities.includes('ssh'))
+                throw new Error('SSH is required for setting up the environment.');
+
+            if ((await this.ssh('stat /etc/apt/sources.list.d/appstraction.sources')).code === 0) return;
```
[the remark this answers is not in the document]
Comment (zner0L) [reply src/ios.ts]: I realized this and threw it out again.

File: src/ios.ts (hunk 1 above)
Comment (zner0L) [reply src/ios.ts:231]: Does `frida`depend on `root`, though? One could install `frida-server` on the device and use it without any need to access the terminal.

Comment (zner0L) [reply src/ios.ts:231]: Otherwise I think you are right. Do you want me to do this in this PR, or open an issue for it?

Comment (baltpeter) [reply src/ios.ts:232]: > Does `frida`depend on `root`, though? One could install `frida-server` on the device and use it without any need to access the terminal.

I don't understand what you're saying. Unless you compile `frida-server` into an app, you can only use it on a jailbroken device, so it does require root?

Comment (baltpeter) [reply src/ios.ts:232]: > Otherwise I think you are right. Do you want me to do this in this PR, or open an issue for it?

Issue is fine for now.

**Comment by baltpeter** [comment]
I tried resetting an rejailbreaking one of my iPhones to test this but [failed](https://github.com/tweaselORG/meta/issues/4#issuecomment-1577102528). As such, I'm really not too keen on also resetting the other phone. I need at least one jailbroken device…

I really don't know how I can test this.

File: src/ios.ts (hunk 1 above)
Comment (zner0L) [reply src/ios.ts:232]: > I don't understand what you're saying. Unless you compile `frida-server` into an app, you can only use it on a jailbroken device, so it does require root?

Yes, but you don’t need to necessarily give appstraction root access, if you installed it yourself.

Comment (baltpeter) [reply src/ios.ts:232]: Hm, I guess you could view it that way.

**Comment by baltpeter** [comment]
Trying the example script on this branch on a freshly reset and jailbroken iOS 15 iPhone (with a [password for `root` set](https://github.com/tweaselORG/appstraction/issues/59#issuecomment-1578612791)), I get the following error:

```
/home/benni/coding/JS/tweasel/appstraction/src/ios.ts:197
                if (!(await retryCondition(fridaIsRunning, 20))) throw new Error('Frida server did not start.');
                                                                       ^


Error: Frida server did not start.
    at Object.ensureFrida (/home/benni/coding/JS/tweasel/appstraction/src/ios.ts:197:72)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at Object.ensureDevice (/home/benni/coding/JS/tweasel/appstraction/src/ios.ts:236:13)
    at <anonymous> (/home/benni/coding/JS/tweasel/appstraction/examples/ios-device.ts:22:5)
```

File: src/index.ts [hunk 3]
```diff
@@ -367,7 +375,7 @@ export type RunTargetOptions<
 export type SupportedCapability<Platform extends SupportedPlatform> = Platform extends 'android'
     ? 'wireguard' | 'root' | 'frida' | 'certificate-pinning-bypass'
     : Platform extends 'ios'
-    ? 'ssh' | 'frida'
+    ? 'ssh' | 'frida' | 'certificate-pinning-bypass'
```
Comment (baltpeter) [inline src/index.ts:378]: We need to document that this will permanently disable cert pinning on the whole device.

**Comment by baltpeter** [comment]
The problem is that I don't have an `/etc/apt` directory that you could write the sources to:

```
Vanessas-iPhone:~ root# ls -l /etc
lrwxr-xr-x 1 root wheel 11 Aug 11  2022 /etc -> private/etc/
Vanessas-iPhone:~ root# ls /private/etc 
asl/  asl.conf  group  hosts  hosts.equiv  master.passwd  networks  notify.conf  passwd  ppp/  protocols  racoon/  services  ttys
```

As such, only the packages that don't need extra repositories installed. And since we're not checking anything for errors, the rest just silently failed.

**Comment by baltpeter** [comment]
So, where are the sources saved?

```sh
Vanessas-iPhone:~ root# grep -r "apt.procurs.us" / 2>/dev/null
/private/preboot/88DCF80B58E42648095610A0AF68DA1FCCCD3DEC60C0E2440FE0CFD6E4B47BD03B96B3EDDB340CB3517674A9059D3429/jb-0N5TRzIB/procursus/etc/apt/sources.list.d/procursus.sources:URIs: https://apt.procurs.us/
/private/preboot/88DCF80B58E42648095610A0AF68DA1FCCCD3DEC60C0E2440FE0CFD6E4B47BD03B96B3EDDB340CB3517674A9059D3429/jb-0N5TRzIB/procursus/var/mobile/Library/Caches/Sileo/RepoHashCache.json:{"sha512":{"https:\/\/apt.procurs.us\/dists\/1800\/Packages.zst":"716ec7a16de7cd9ebd9d5c3936e16cf8902a4fd0f0e2fc5efdc75bc82db4d330bd9ee474e034f227d13736acd9cd8509ab23fb94d749b9 […long]
/private/var/mobile/Library/Caches/org.coolstar.SileoStore/fsCachedData/3D1F8351-0668-411C-9D23-22FBD58EC88B:{"status":"200 OK","date":"2 [very, very, long…]
/private/var/mobile/Library/palera1n/logs/1686050641.916493-loader.log:Get:2 https://apt.procurs.us 1800 InRelease [8310 B]
/private/var/mobile/Library/palera1n/logs/1686050641.916493-loader.log:Get:5 https://apt.procurs.us 1800/main iphoneos-arm64 Packages [308 kB]
/private/var/mobile/Library/palera1n/logs/1686050641.916493-loader.log:Get:1 https://apt.procurs.us 1800/main iphoneos-arm64 libkrw0-tfp0 iphoneos-arm64 1.1.1 [3862 B]
/private/var/mobile/Library/palera1n/logs/1686050641.916493-loader.log:Get:1 https://apt.procurs.us 1800/main iphoneos-arm64 nebula-keyring all 2022.09.15 [3608 B]
```

The first one is the only one that makes sense. But it's in a `procursus` directory. Did I just happen to pick the one source with the same name as the thing that holds these files or does every source have its own… whatever?

I added a dummy source:

```sh
Vanessas-iPhone:~ root# grep -r "test.bn.al" / 2>/dev/null
/private/preboot/88DCF80B58E42648095610A0AF68DA1FCCCD3DEC60C0E2440FE0CFD6E4B47BD03B96B3EDDB340CB3517674A9059D3429/jb-0N5TRzIB/procursus/etc/apt/sources.list.d/sileo.sources:URIs: https://test.bn.al/
/private/var/tmp/488DC191-B899-470F-A7C6-62CA3503FE70:URIs: https://test.bn.al/

Vanessas-iPhone:~ root# cat /private/preboot/88DCF80B58E42648095610A0AF68DA1FCCCD3DEC60C0E2440FE0CFD6E4B47BD03B96B3EDDB340CB3517674A9059D3429/jb-0N5TRzIB/procursus/etc/apt/sources.list.d/sileo.sources
Types: deb
URIs: https://havoc.app/
Suites: ./
Components:

Types: deb
URIs: https://repo.chariz.com/
Suites: ./
Components:

Types: deb
URIs: https://test.bn.al/
Suites: ./
Components:
```

Okay, `/private/preboot/88DCF80B58E42648095610A0AF68DA1FCCCD3DEC60C0E2440FE0CFD6E4B47BD03B96B3EDDB340CB3517674A9059D3429/jb-0N5TRzIB/procursus/etc/apt` it is.

**Comment by baltpeter** [comment]
Hm, I can't find the `re.frida.server.plist` you're trying to modify:

```sh
Vanessas-iPhone:~ root# cat /Library/LaunchDaemons/re.frida.server.plist
cat: /Library/LaunchDaemons/re.frida.server.plist: No such file or directory
Vanessas-iPhone:~ root# ls /Library/LaunchDaemons 
com.apple.factory.NFQRCoded.plist
Vanessas-iPhone:~ root# ls /private/preboot/88DCF80B58E42648095610A0AF68DA1FCCCD3DEC60C0E2440FE0CFD6E4B47BD03B96B3EDDB340CB3517674A9059D3429/jb-0N5TRzIB/procursus/Library/LaunchDaemons/
com.apple.atrun.plist  com.openssh.sshd.plist  us.diatr.shshd.plist
Vanessas-iPhone:~ root# find / -name "re.frida.server.plist" 2>/dev/null
```

**Comment by baltpeter** [comment]
I'm assuming you installed in rootful mode? Maybe we should just make that a requirement and document it in the README?

File: src/index.ts (hunk 3 above)
Comment (zner0L) [reply src/index.ts:378]: Yes, but what would be the place for that?

Comment (baltpeter) [reply src/index.ts:378]: We should start documenting what the capabilities mean, anyway. Either in `SupportedCapability` or `PlatformApiOptions`.

**Comment by zner0L** [comment]
Yes. I use rootful mode, which I think we should require.

**Comment by baltpeter** [comment]
Please document your steps for jailbreaking. I'm really not sure which of the options I would use for that.

**Comment by zner0L** [comment]
> Please document your steps for jailbreaking. I'm really not sure which of the options I would use for that.

https://github.com/tweaselORG/meta/issues/4#issuecomment-1578874305

**Comment by baltpeter** [comment]
With rootful mode, this now works on my iOS 15 device. Unfortunately, SSL Kill Switch 2 doesn't seem to be enabled after the automatic setup. The menu item doesn't show up in the Settings app. Iirc, you need to respring for that to start working. I tried `launchctl reboot userspace` but that didn't help.

**Comment by baltpeter** [comment]
Huh. Even after respringing, the setting still doesn't show up. Maybe the version of SKS2 you're installing doesn't have that? Have you verified whether it works? I can't think of a good app to test right now.

**Comment by baltpeter** [comment]
Other than that, it did work. I was able to run the the example successfully (well, the app I tested crashed on the first try, causing the script to fail—but it did work on second try…).

**Comment by zner0L** [comment]
Ok, I tested it and I can confirm that SSL Kill Switch 2 is activated. However I forgot to include `PreferenceLoader` as a dependency for it to show up in the Preferences. This needs more than a respring, you need to go into the palera1n app and tap "Enable Tweaks". I haven’t figured out what that does, yet.

**Comment by zner0L** [comment]
> For future reference: How?

SSL Kill Switch writes to the syslog: `=== SSL Kill Switch 2: Preference set to 1.`. Also, the App Store relies on certificate pinning and crashes if mitmproxy is used as a proxy, even with the CA set up. But it does work if SSL Kill Switch 2 has been installed, even without respring.

> Respringing unfortunately leaves the device locked. In the current state, the example now fails with `Error: Cannot connect using Frida.` while it is still respringing.

This is probably because the SpringBoard has not fully started up again. I guess we should `waitForDevice` there, but this is not available in the scope of `respring`… Also, for future reference, there is a way to unlock the screen: https://stackoverflow.com/questions/21706050/how-to-unlock-ios-screen-programmatically

> Is the problem only the preference or does the pinning bypass also not work currently? I don't care too much about the former.

The pinning bypass also seems to work without `PreferenceLoader` and without respring. So I guess if you want to marge you could undo the last changes I made.

**Comment by baltpeter** [comment]
> So I guess if you want to marge you could undo the last changes I made.

Okay, will do.

**Comment by baltpeter** [comment]
> I was able to run the the example successfully (well, the app I tested crashed on the first try, causing the script to fail—but it did work on second try…).

Tested again on a fresh install and ran into the same problem, regardless of which app I tested. Even manually starting the installed apps afterwards, they would always crash.

I found that I have to start the apps once _without_ the proxy disabled. Afterwards, they also work with the proxy, even after uninstalling and reinstalling them.

This seems like it might be a cert pinning issue with Apple verifying that I do in fact own the app? I don't remember running into this before.

**Comment by baltpeter** [comment]
Interestingly, it now also works with a different app without needing to start it once without a proxy. So maybe, the condition is in fact that I need to start _any_ App Store app once without proxy? Either way, that doesn't seem so severe anymore and doesn't seem to be caused by this PR, so I'll push this into an issue and merge.

**Review by baltpeter — APPROVED** [review]

**Pull request merged**